### DISCUSSION TITLE
Docs: How to simulate an app/global-error.js | tsx

### DISCUSSION THREAD

**Discussion by icazemier** [vercel/next.js#73293] [Help]
### What is the improvement or update you wish to see?

Would be nice, to somehow trigger an error which shows this page

### Is there any context that might help us understand?

For testing purposes

### Does the docs page already exist? Please link to it.

https://beta.nextjs.org/docs/api-reference/file-conventions/error#global-errorjs

**Comment by yannbolliger** [comment]
You can simply throw an error in your root/top-level `layout.tsx` and launch your app with `next start` (in dev you'll see the dev error view).

**Comment by icazemier** [comment]
> You can simply throw an error in your root/top-level `layout.tsx` and launch your app with `next start` (in dev you'll see the dev error view).

Dear @yannbolliger ,

I tried something like:

```jsx
export default function RootLayout({ children }: { children: React.ReactNode }) {
// 🙏🏼
throw new Error('Whoops');

  return (
    <html lang="en" translate="no">
....
```

in : `src/app/layout.tsx` (appDir:true)

or even with a setTimeout around it, but I just can't seem to simulate a situation where I see the global-error page popping up.

Could you give me an example?

**Comment by RafalChojnowskiPagepro** [comment]
This example fails for me too

**Comment by yannbolliger** [comment]
Did you use the production build `next start`?

**Comment by icazemier** [comment]
Hi NextJS team,

I'm still struggling to "really" see the global-error page live in action.

Seems related to this (`locked and limited conversation to collaborators`) issue: https://github.com/vercel/next.js/issues/46572

I can reproduce the fact global-error.tsx is (still) not working with the following steps to reproduce:
(Pushed here: https://github.com/icazemier/nextjs-global-error-issue)

# Install nextjs (corresponding to the docs)
![image](https://github.com/vercel/next.js/assets/64091093/73b634b0-394b-4d6f-9f5b-96b39c416105)

# Create `global-error.tsx` in `/app` (as documented)

```typescript
"use client";

export default function GlobalError({
  error,
  reset,
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  return (
    <html>
      <body>
        <h2>Something went wrong!</h2>
        <button onClick={() => reset()}>Try again</button>
      </body>
    </html>
  );
}

```


# Throw a async error in the root layout file

```typescript

import "./globals.css";
import type { Metadata } from "next";
import { Inter } from "next/font/google";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  // Throw an error here!
  setTimeout(() => {
    throw new Error("Whoops");
  }, 2000);

  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
    </html>
  );
}

```

# Build and run the project (production variant)

`npm run build`

![image](https://github.com/vercel/next.js/assets/64091093/69b84936-c74b-4f83-a626-02fcec0bfc43)

`npm start`

## Navigate to the root page in the browser and wait for the error

## Front end doesn't give a clue (no global-error triggererd)
![image](https://github.com/vercel/next.js/assets/64091093/61fe0857-f245-4fb9-b693-402d48f1880a)

## Terminal doesn't show anything

![image](https://github.com/vercel/next.js/assets/64091093/70e85f30-3411-465a-a7e0-74b31b098520)


# Build and run the project in development mode

Nothing in the front end either, but in the terminal it does show the error??

![image](https://github.com/vercel/next.js/assets/64091093/418b20fb-a9f4-429a-95f3-b70638cfd856)


Obviously there is something wrong, because no global error is triggered.

Or I'm doing weird things (which often is the case .... 😅).

But my ticket/request still stands.

**Comment by icazemier** [comment]
Not sure, but seeing several related issues, possibly...
Tagging:
- https://github.com/vercel/next.js/issues/53756
- https://github.com/vercel/next.js/issues/55647
- https://github.com/vercel/next.js/issues/52993
- https://github.com/vercel/next.js/issues/50119
- https://stackoverflow.com/questions/76827088/global-error-tsx-dont-catch-error-in-next-js

**Comment by icazemier** [comment]
In the meantime I managed to get the error boundary working!

Partially because I didn't understand why the boundaries where not catching them...
I read about it here: https://semaphoreci.com/blog/error-handling-layer-react#limitations-of-this-approach

Made a error catch provider to at least trigger the error boundary and showing itself

See: https://github.com/icazemier/nextjs-global-error-issue/tree/22da1965208eae721aafa3ca8e2ef80ad5cdc468

(Still not able to simulate the global-error.tsx)

**Comment by ADTC** [comment]
# How to simulate an app/global-error.js | tsx

Things you must understand first:

* It can only be simulated using a production build. You cannot run a development server to simulate it because the dev server will show the standard development mode error message with stack trace. _(There's no need to deploy to Vercel. You can run the production build locally.)_
* The error must be thrown dynamically, because any statically thrown error will be detected during build process as "always thrown", and the build will fail.
* Styles imported in `global-error.*` from CSS modules will not apply unless they are also imported in `layout.*` and somehow used there too. This is because NextJS doesn't download the styles imported in `global-error.*` to the client even though it uses the class names on the elements. _(This could be a bug with Next.js 14.)_
  * A clean way to circumvent this is to use global styles instead of importing a module. But if you really need to import modules, the hack to achieve this is mentioned below.
* The `html` and `body` tags should be present in `global-error.*`. If you have custom attributes on them in `layout.*` you may need to copy them. (This depends on whether the attributes are required to properly render the page.)
* You cannot use the `useRouter` hook in `global-error.*`. If you want to have a button that refreshes the page, you should use plain JavaScript: `onClick={() => window.location.reload()}`.

## Setting up:

Create a `global-error.tsx` file as per the docs and the above tips.

### Button to reload page in the `global-error.tsx` file:

Skip this if you don't need a Reload button. But if you do, since we cannot use the `useRouter` hook here, we depend on regular Javascript to do a full-page refresh:

```tsx
<button onClick={() => window.location.reload()}>Reload page</button>
```

Note: The `reset: () => void` parameter doesn't work for this. It just throws weird errors.

### Fixing the issue of styles not getting applied in `global-error.tsx`:

Skip this if you don't have any CSS modules imported in `global-error.tsx`. But if you do, for example:

```tsx
import Styles from '@/styles/error.module.scss'
```

**The cleanest and eaiest way is to use global styles instead of importing a module.**

For example, you may have a global styles file in `app` called `globals.scss` and you are already importing it into `layout.tsx`:

```tsx
import './globals.scss'
```

You can create unique class names in this file with a globally unique prefix like `ErrorPage__` then use them in `global-error.tsx`.

The other hacky way if you want to import a module in `global-error.tsx` is to add a dummy class in the module:

```scss
// A dummy element to force load the styles in layout.tsx
.dummy {
  display: none;
}
```

And also import the module into `layout.tsx`:

```tsx
import ErrorStyles from '@/styles/error.module.scss'
```

And use it in a `div` somewhere inside the `body` tag (it doesn't matter where) in the `layout.tsx` file:

```tsx
<div className={ErrorStyles.dummy} />
```

## Testing:

The following code is for testing the global error page. It should not be deployed in an actual production deployment.

In order to test this, a component should be dynamically rendered at runtime. Any statically generated component will be pre-built at build time and therefore fail the build if an error is thrown.

One easy way I managed to test this is in a client component that conditionally adds something:

```tsx
{condition && <Component />}
```

In the client component file, I add a new function:

```tsx
const fakeError = () => {
  console.log('Throwing a fake error!')
  throw new Error('Whoops!')
}
```

Then I use this as a className attribute value. (It doesn't matter what you use it as. I chose className because it didn't throw a type error.)

```tsx
{condition && (
  <>
    <Component />
    <div className={fakeError()} />
  </>
)}
```

Now I build and run the production build, using `npm` or `next`:

```sh
npm run build && npm run start
# OR:
next build && next start
```

Navigate on `localhost` to the page which will fulfill the `condition`. And viola! _(This is my custom error page. Your own style will definitely be different.)_

<img width="912" alt="image" src="https://github.com/vercel/next.js/assets/6047296/195296ff-3958-488f-9bb3-e7258b885523">

**Tip:** The `error` object has some properties like `name`, `message` and `digest` which you can use to display the error details on the `global-error` page.

_Good luck!_

**Comment by psirenny** [comment]
This is very very bad DX.

**Comment by ADTC** [comment]
Apparently you _can_ simulate it in dev by making a page route which has a client component which calls `useRouter()` and has a button that calls `router.push` to another route which is a server component that calls `redirect` from `next/navigation` to a third route. For some reason Next.js doesn't like this and throws a "Rendered more hooks than during the previous render" error. This triggers the `global-error` page where the error could be displayed.

**Comment by Antonio-Laguna** [comment]
I have created this component. Include it on any page:

```tsx
'use client';

import { useState, useEffect } from 'react';

export const BrokenModule = (props: unknown) => {
	const [isBroken, setIsBroken] = useState(false);

	useEffect(() => {
		setTimeout(() => {
			setIsBroken(true);
		}, 2000);
	}, []);

	if (isBroken) {
		const test = props.test.not;
		return <p>{test}</p>;
	}

	return <p>Broken module</p>;
}
```

Guaranteed to get an error after 2 seconds.

**Comment by ADTC** [comment]
@Antonio-Laguna unfortunately your method can only test `error.tsx` but not `global-error.tsx`. I noticed this because I have different contents for each of these files.

As of now, to test the `global-error.js | tsx` file, the best known method is to do what I did and do a production build:

```tsx
const fakeError = () => {
  console.log('Throwing a fake error!')
  throw new Error('Whoops!')
}

export default function FakeError({ condition }: { condition: boolean }) {
  return condition && <div className={fakeError()} />
}
```

Somewhere in your code, like `layout.tsx`:

```tsx
<FakeError condition={true} />
```

As a reminder, you have to use `build` to create a production build and serve it. In NextJS projects, this is usually done with `npm run serve`. For any changes, you'll have to kill the server and rebuild a new production build.

**Comment by mihaildono** [comment]
I cant manage to catch the error...  This is my `app/error.tsx`
``` 'use client';

import { useEffect } from 'react';

import { logException } from '@lib/utils';

export default function Error({
  error
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  useEffect(() => {
    console.error('error');
    logException(error);
  }, [error]);

  return (
    <div>
      <h2>Something went wrong!Error</h2>
    </div>
  );
}
```

This is my `app/global-error.tsx`


``` 'use client';

import { useEffect } from 'react';

import { logException } from '@lib/utils';

export default function GlobalError({
  error
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  useEffect(() => {
    console.error('global error');
    logException(error);
  }, [error]);

  return (
    <html>
      <body>
        <h2>Something went wrong!Global</h2>
      </body>
    </html>
  );
}
```

Set this `ENV NODE_ENV production` in my docker file and start my server with docker - so it should be prod?

How I trigger the error in a client component 

``` 
<Button
          onClick={() => {
            throw new Error('custom err');
          }}
        />
```

I see the error thrown in console from client component button, but not in the error components.

` "next": "14.0.4" ` ` "react": "^18", ` This was the default I initialised the project with

**Comment by keighl** [comment]
How I was able to view my `global-error.tsx`

```tsx
// app/page.tsx

// Don't let next cache a functioning index.html at build time
export const dynamic = "force-dynamic";

export default async function Homepage() { 
  // ...
}
```

```tsx
// app/layout.tsx

const RootLayout = async ({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) => {
  
  if (process.env.NEXT_PHASE !== PHASE_PRODUCTION_BUILD) {
    // This will allow the build, but fail at runtime
    throw new Error('Oh nooo!');
  }

  return <html /> 
}
```

**Comment by aryomuzakki** [comment]
I got here even though I managed to simulate it before with React Developer Tools and I answered the question in here: [https://stackoverflow.com/a/78671438/12479141](https://stackoverflow.com/a/78671438/12479141)

Here is my solution to test `global-error.jsx` in development mode:
1. install [React Developer Tools](https://react.dev/learn/react-developer-tools)
2. open your browser Developer Tools, and open the new 'Component' tab
3. select the component inside 'Error Boundary' component, and click the _exclamation mark_ (!) icon _(as you can see in the screenshot)_
_[UPDATE] ⚠ **IMPORTANT!** Make sure to just choose 'Router' component instead, since deeper component after 'HotReload' component will fail to show `global-error.jsx` file_

![gwwYDAHI](https://github.com/user-attachments/assets/98901932-2de7-4c16-99b1-c90b21ad304a)

[UPDATE] 
Things to note:
- Hot Reload doesn't work, reload manually and retrigger component error from dev tools to see changes in `global-error.jsx` file
- Any error (like undeclared variable or missing import) in `global-error.jsx` file will not show. So try to test it in `error.jsx` file, then copy and paste content. `<html>` and `<body>` tag still need to be added in `global-error.jsx`
- These steps won't work in production, instead, throw an error in layout file (not page file) like @ADTC solution
- Tested in Next 14.x and Next 15.0.3 + React 19 RC

Simulated error example in a small project :
[https://github.com/aryomuzakki/nextjs-error-simulated-example](https://github.com/aryomuzakki/nextjs-error-simulated-example)

**Comment by theogravity** [comment]
@aryomuzakki I see the option in dev mode, but it doesn't get me to the global error page. I don't see the option in prod mode. How are you getting it to work in dev mode? Using Next 15 RC + React 19 RC.

**Comment by firelizzard18** [comment]
@ADTC is it feasible to add an option that disables the normal development error page?

**Comment by ADTC** [comment]
@firelizzard18 what do you mean? Development error pages only show in development builds. They won't show in production builds. If you want to test the app using a production build, you can do so using `npm run build` or `npm run serve`. Note that this will not hot-reload any code changes, so you'll have to kill and restart the production build server.

💡 If you're just trying to distinguish between `error` and `global-error`, it's easier to simply have different contents for each of those files. For example, one could say "This is the regular error page." while the other says "This is the global error page." with some different background colors, maybe.

**Comment by AleferReinert** [comment]
> Cheguei aqui mesmo tendo conseguido simular antes com o React Developer Tools e respondi a pergunta aqui: https://stackoverflow.com/a/78671438/12479141
> 
> Aqui está minha solução:
> 
> 1. instalar [React Developer Tools](https://react.dev/learn/react-developer-tools)
> 2. abra as Ferramentas do desenvolvedor do seu navegador e abra a nova aba 'Componente'
> 3. selecione o componente dentro do componente 'Error Boundary' e clique neste ícone (!)
>    ![gwwYDAHI](https://private-user-images.githubusercontent.com/66844553/365701885-98901932-2de7-4c16-99b1-c90b21ad304a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxMjEwMjIsIm5iZiI6MTczMjEyMDcyMiwicGF0aCI6Ii82Njg0NDU1My8zNjU3MDE4ODUtOTg5MDE5MzItMmRlNy00YzE2LTk5YjEtYzkwYjIxYWQzMDRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDE2Mzg0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzMmIyMGExMjA0N2ZiNWQ0MmQ3ZTkyZjBjYjg3ZDkwZTRhYzJkMTE1ZTg3MGMzNzZmNWUzMDQ0Y2UyNTNjMjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hpijH6aRU4umIm_I0iStxZG7jywUzQlo5ng6m64XWMY)

Works for me, thanks!

**Comment by firelizzard18** [comment]
> @firelizzard18 what do you mean? Development error pages only show in development builds. 

@ADTC I mean a flag for `next dev` that allows me to test my error pages. Having to rebuild the production build every time I make a change is rather cumbersome. If I could run something like `next dev --disable-dev-error-page` then I could test my error pages much more easily.

**Reply by ADTC** [reply]
It's true that a flag like that would be helpful. Perhaps it could be filed as a feature request.

**Comment by aryomuzakki** [comment]
@theogravity i update some notes to my solution, I tested it and it works on Next 15.0.3 + React 19 RC, but not tried on Next Canary. I also add a project example.

@firelizzard18 you can check my solution to test your error pages in development 

@ADTC but still, there should be a better way to do this

**Reply by ADTC** [reply]
It's true, there could be a better way. A flag like `--disable-dev-error-page` would be helpful. Perhaps it could be filed as a feature request.
